### PR TITLE
Refactor the beta codebase to solve `mypy` issues in integration test usage

### PR DESCRIPTION
**Pull request by tsmith023**
This PR makes a number of changes that fix mypy errors and warnings within the integration tests (and therefore the actual usage of the API). At a high level, the key issues hinted at were:

- Poorly defined input types due to invariance vs covariance confusion
  - Biggest offenders were `Dict` and `List` in function arguments which are in-fact invariant types within the python typing system. These have been changed to `Mapping` and `Sequence` respectively where appropriate
 
- Since splitting `Properties` and `References` out, it also became apparent to split the behaviour of `_Reference` also to aid in the generics flow
  - `_Reference` is now only relevant to insert data, `_CrossReference` is only relevant to fetching data. `_CrossReference` has `.objects` while `_Reference` has `.uuids` and `.target_collection`
  - The motivation was that the actual data type/schema of the referenced collection is not relevant when inserting it into collection so it shouldn't be required at all when the user does so

- Finally, this branch was checked-out from this branch: https://github.com/weaviate/weaviate-python-client/pull/659, so that PR should be merged first before this one is reviewed

### CODE REVIEW

File: weaviate/collections/classes/data.py [hunk 1]
```diff
@@ -21,14 +20,20 @@ class RefError:
...
     """This class represents an entire object within a collection to be used when batching."""
 
-    properties: Optional[P] = None
+    properties: P
     uuid: Optional[UUID] = None
     vector: Optional[List[float]] = None
-    references: Optional[R] = None
+    references: R = None  # type: ignore
+    # R is clearly bounded to Optional[Any] and defaults to None but mypy doesn't seem to understand that
+    # throws error: Incompatible types in assignment (expression has type "None", variable has type "R")  [assignment]
```
Comment (dirkkul) [inline weaviate/collections/classes/data.py:36]: make a bugreport?

File: weaviate/collections/classes/data.py [hunk 2]
```diff
@@ -21,14 +20,20 @@ class RefError:
...
+P = TypeVar("P")
+R = TypeVar("R", bound=Optional[Any], default=None)
+
+
 @dataclass
 class DataObject(Generic[P, R]):
     """This class represents an entire object within a collection to be used when batching."""
 
-    properties: Optional[P] = None
+    properties: P
```
Comment (dirkkul) [inline weaviate/collections/classes/data.py:31]: why this change?

File: weaviate/collections/classes/internal.py [hunk 3]
```diff
@@ -284,13 +276,28 @@ def target_collection(self) -> str:
         """Returns the target collection name."""
         return self.__target_collection
 
+
+class _CrossReference(Generic[Properties, IReferences]):
+    def __init__(
+        self,
+        objects: Optional[List[_Object[Properties, IReferences]]],
+    ):
+        """You should not initialise this class directly. Use the `.to()` or `.to_multi_target()` class methods instead."""
```
Comment (dirkkul) [inline weaviate/collections/classes/internal.py:285]: docstring does not fit

File: weaviate/collections/classes/internal.py [hunk 4]
```diff
@@ -352,15 +357,18 @@ def to_multi_target(
...
+        return _Reference(
             target_collection.name
             if isinstance(target_collection, _CollectionBase)
             else target_collection,
             uuids,
         )
 
 
+WeaviateReference: TypeAlias = _Reference
+WeaviateReferences: TypeAlias = Mapping[str, WeaviateReference]
```
Comment (dirkkul) [inline weaviate/collections/classes/internal.py:369]: I have the feeling we have to many classes that kind of sound the same, but have different usages and it is not obvious which one is used where (just thinking of all the references) - maybe we can make this more somehow obvious 🤔

File: weaviate/collections/collection.py [hunk 5]
```diff
@@ -268,8 +268,10 @@ def iterator(
...
                     include_vector=include_vector,
                     return_metadata=return_metadata,
-                    return_properties=return_properties or self.__properties,
-                    return_references=return_references,
+                    return_properties=return_properties,  # type: ignore
+                    return_references=return_references,  # type: ignore
                 ).objects
             ),
+            # type ignores here are because we don't care about the types when working within the API
+            # types are mainly for the users of the API
```
Comment (dirkkul) [inline weaviate/collections/collection.py:276]: why do the types not fit here?

File: weaviate/collections/classes/internal.py (hunk 4 above)
Comment (tsmith023) [reply weaviate/collections/classes/internal.py:369]: Will resolve in separate PR

File: weaviate/collections/classes/data.py (hunk 1 above)
Comment (tsmith023) [reply weaviate/collections/classes/data.py:36]: Will do!
